### ISSUE TITLE
Create a "routing" label placement mode

### ISSUE THREAD

**Issue by lucaswoj** [mapbox/mapbox-gl-js#3992]
... or perhaps create a custom label placement plugin API

> @ansis:
> Dynamically adapting what labels are shown and how they look based on a route.
> 
> For example:
> - making roads used by the route have bigger labels
> - deemphasizing roads that don't cross the route
> - placing labels for intersecting roads close to the intersection
> - adding pois near turns to make them easier to identify
> 
> I don't think this would be entirely implemented within core mapbox gl but we would need to implement some apis to make this possible.

**Comment by 1ec5** [comment]
Is this related to https://www.mapbox.com/blog/labels-for-navigation/?

**Comment by lucaswoj** [comment]
@1ec5 Yes. That blog post demos a prototype implementation of this feature.

**Comment by 1ec5** [comment]
In that case, these two branches appear to be prototypes of the feature:

https://github.com/mapbox/mapbox-gl-js/tree/guidance-labelling
https://github.com/mapbox/mapbox-gl-js/tree/guidance-labelbox-anglefade

**Comment by natslaughter** [comment]
@lucaswoj for the [turn-by-turn navigation map](https://github.com/mapbox/cartography/issues/90), I'm very interested in learning more about road label placing and styling possibilities, as they relate to a route-line. Is there a current schedule for implementing the ability to style road labels along a route differently from other road labels? Would it be realistic to develop the capabilities you listed above so they could be a part of the [navigation map](https://github.com/mapbox/cartography/issues/85) we are shipping the first week in March? 

@nickidlugash @pveugen @aparlato
